pkgs/servers/sql/postgresql/generic.nix

```diff
@@ -0,0 +1,320 @@
+              rm "$i"
+            fi
+          done
+        fi
+      '' + lib.optionalString jitSupport ''
+        # Move the bitcode and libllvmjit.so library out of $lib; otherwise, every client that
+        # depends on libpq.so will also have libLLVM.so in its closure too, bloating it
+        moveToOutput "lib/bitcode" "$out"
+        moveToOutput "lib/llvmjit*" "$out"
+
+        # In the case of JIT support, prevent a retained dependency on clang-wrapper
+        substituteInPlace "$out/lib/pgxs/src/Makefile.global" --replace ${stdenv'.cc}/bin/clang clang
+        nuke-refs $out/lib/llvmjit_types.bc $(find $out/lib/bitcode -type f)
+
+        # Stop out depending on the default output of llvm
+        substituteInPlace $out/lib/pgxs/src/Makefile.global \
+          --replace ${llvmPackages.llvm.out}/bin "" \
+          --replace '$(LLVM_BINPATH)/' ""
+
+        # Stop out depending on the -dev output of llvm
+        substituteInPlace $out/lib/pgxs/src/Makefile.global \
+          --replace ${llvmPackages.llvm.dev}/bin/llvm-config llvm-config \
+          --replace -I${llvmPackages.llvm.dev}/include ""
+
+        ${lib.optionalString (!stdenv'.isDarwin) ''
+          # Stop lib depending on the -dev output of llvm
+          rpath=$(patchelf --print-rpath $out/lib/llvmjit.so)
+          nuke-refs -e $out $out/lib/llvmjit.so
+          # Restore the correct rpath
+          patchelf $out/lib/llvmjit.so --set-rpath "$rpath"
+        ''}
+      '';
+
+    postFixup = lib.optionalString (!stdenv'.isDarwin && stdenv'.hostPlatform.libc == "glibc")
+      ''
+        # initdb needs access to "locale" command from glibc.
+        wrapProgram $out/bin/initdb --prefix PATH ":" ${glibc.bin}/bin
+      '';
+
+    doCheck = !stdenv'.isDarwin;
+    # autodetection doesn't seem to able to find this, but it's there.
+    checkTarget = "check";
+
+    preCheck =
+      # On musl, comment skip the following tests, because they break due to
+      #     ! ERROR:  could not load library "/build/postgresql-11.5/tmp_install/nix/store/...-postgresql-11.5-lib/lib/libpqwalreceiver.so": Error loading shared library libpq.so.5: No such file or directory (needed by /build/postgresql-11.5/tmp_install/nix/store/...-postgresql-11.5-lib/lib/libpqwalreceiver.so)
+      # See also here:
+      #     https://git.alpinelinux.org/aports/tree/main/postgresql/disable-broken-tests.patch?id=6d7d32c12e073a57a9e5946e55f4c1fbb68bd442
+      if stdenv'.hostPlatform.isMusl then ''
+        substituteInPlace src/test/regress/parallel_schedule \
+          --replace "subscription" "" \
+          --replace "object_address" ""
+      '' else null;
+
+    doInstallCheck = false; # needs a running daemon?
+
+    disallowedReferences = [ stdenv'.cc ];
+
+    passthru = let
+      this = self.callPackage generic args;
+      jitToggle = this.override {
+        jitSupport = !jitSupport;
+      };
+    in
+    {
+      psqlSchema = lib.versions.major version;
```

Comment:
I'm wondering if we should still expose `jitSupport` here, that's the easiest way for other packages to check if postgresql is built with or without that.

Reply:
So, the way I thought about this was: we effectively have two variants of postgresql packaged, JIT and non-JIT (and multiple releases, but those can be found out via `.version`) and I think it doesn't hurt to expose the (rather prominent) detail of the package.

One thing I could imagine is a reusable module that expects to have a JIT variant for $service and asserts against that. It's not the reponsibility of that module to set the postresql package if it's for a different service and I don't think you want that because you'll run into issues quickly then when you'll need to patch the package on your own.

Reply:
> But JIT support doesn't change PostgreSQL functionally, afaik

To my knowledge, yes. But modules warning in case of unsuitable configurations is still a legit use-case (and I can certainly imagine applications where you want that).

> Personally, I could see more benefit in exposing enableSystemd or enablePython

I think you're right.
Postgresql extensions have a nice way to discover that and if you really need it you can still use `hasInfix "-jit-"`.

The core problem behind this is that we just don't have a good configuration story for packages (as in, you can read from `config.services.foo` how service `foo` is configured, but you can't do the same for `pkgs.foo` out of the box), so this should be solved at a different level.

> btw, just noting: Should we maybe rename jitSupport to enableJIT for consistency

Yes, but I'd suggest to leave `jitSupport ? null` around for 1 release and raise a warning if it's not null (i.e. `true`/`false`). `enableJIT` should be set to `jitSupport` in that case, then. Sure, NixOS users won't even notice because they'll probably use the module, but I'm pretty sure there are a few nixpkgs users out there.

These package arguments not only have horrible discovery, but as a result also horrible (i.e. non-existant) ways to communicate changes (short of "This attribute isn't supposed to be part of this function argument").